torch_geometric/transforms/rooted_subgraphs.py

```diff
@@ -0,0 +1,230 @@
+from typing import Optional, Tuple
+
+import torch
+from torch_sparse import SparseTensor
+
+from torch_geometric.data import Data
+from torch_geometric.transforms import BaseTransform
+
+
+class RootedSubgraphData(Data):
+    r"""A data object describing a homogeneous graph together with each node's
+    rooted subgraph. It contains several additional properties that hold the
+    information of all nodes' rooted subgraphs. Assume the data represents a
+    graph with :math:`N` nodes and :math:`M` edges, also each node
+    :math:`i\in \[N\]` has a rooted subgraph with :math:`N_i` nodes and
+    :math:`M_i` edges.
+
+    Additional Properties:
+        subgraph_nodes_mapper (LongTensor): map each node in rooted subgraph
+        to a node in the original graph.
+            Size: :math:`\sum_{i=1}^{N}N_i x 1`
+        subgraph_edges_mapper (LongTensor): map each edge in rooted subgraph to
+         a edge in the original graph.
+            Size: :math:`\sum_{i=1}^{N}M_i x 1`
+        subgraph_batch: map each node in rooted subgraph to its corresponding
+        rooted subgraph index.
+            Size: :math:`\sum_{i=1}^{N}N_i x 1`
+        combined_rooted_subgraph: edge_index of a giant graph which represents
+        a stacking of all rooted subgraphs.
+            Size: :math:`2 x \sum_{i=1}^{N}M_i`
+
+    The class works as a wrapper for the data with these properties, and
+    automatically handles mini batching for them.
+
+    """
+    def __inc__(self, key, value, *args, **kwargs):
+        num_nodes = self.num_nodes
+        num_edges = self.edge_index.size(-1)
+        if 'combined_subgraph' in key:
+            return getattr(
+                self, key[:-len('combined_subgraph')] +
+                'subgraph_nodes_mapper').size(0)
+        elif 'subgraph_batch' in key:
+            # should use number of subgraphs or number of supernodes.
+            return 1 + getattr(self, key)[-1]
+        elif 'nodes_mapper' in key or 'selected_supernodes' in key:
+            return num_nodes
+        elif 'edges_mapper' in key:
+            return num_edges
+        else:
+            return super().__inc__(key, value, *args, **kwargs)
+
+    def __cat_dim__(self, key, value, *args, **kwargs):
+        if 'combined_subgraph' in key:
```

Comment:
```suggestion
        if 'combined_subgraph' in key:
```
Shall we name it something like `*_edge_index` as well?